### PR TITLE
Migrate ProgramHeartbeatDataset to the new SPI

### PR DESCRIPTION
**Pull request by jayfjin**
Migrate ProgramHeartbeatDataset to the new SPI

Builds:
https://builds.cask.co/browse/CDAP-DUT6851-2
https://builds.cask.co/browse/IT-ITM-110

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 1]
```diff
@@ -769,4 +772,36 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
       tableAdmin.create(RUNTIMES_SPEC);
     }
   }
+
+  /**
+   * Schema for program heartbeat
```
Comment (CuriousVini) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: nit: period at the end

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 2]
```diff
@@ -769,4 +772,36 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
...
+
+    public static final String NAMESPACE_FIELD = "namespace";
+    public static final String TIMESTAMP_FIELD = "timestamp";
+    public static final String APPLICATION_FIELD = "application";
+    public static final String PROGRAM_TYPE_FIELD = "program_type";
+    public static final String PROGRAM_FIELD = "program";
+    public static final String RUN_FIELD = "run";
+    public static final String RUN_RECORD = "run_record";
+
+    public static final StructuredTableSpecification RUNTIMES_SPEC = new StructuredTableSpecification.Builder()
```
Comment (CuriousVini) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: why name it `RUNTIMES_SPEC `?

File: cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java [hunk 3]
```diff
@@ -1,193 +0,0 @@
...
-
-  /**
-   * Static method for creating an instance of {@link ProgramHeartbeatDataset}.
-   */
-  public static ProgramHeartbeatDataset getOrCreate(DatasetContext datasetContext,
-                                                    DatasetFramework datasetFramework, CConfiguration cConfiguration) {
-    try {
-      TableProperties.Builder props = TableProperties.builder();
-      long ttl = cConfiguration.getLong(Constants.ProgramHeartbeat.HEARTBEAT_TABLE_TTL_SECONDS);
-      props.setTTL(ttl);
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java:78]: Is there a way to set ttl in new spis? if not, how the clean up will happen?

File: cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatTable.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+public class ProgramHeartbeatTable {
+  private static final Gson GSON = TriggeringScheduleInfoAdapter.addTypeAdapters(new GsonBuilder()).create();
+  private final StructuredTable table;
+
+  public ProgramHeartbeatTable(StructuredTableContext context) {
+    this.table = context.getTable(StoreDefinition.ProgramHeartbeatStore.PROGRAM_HEARTBEATS);
+  }
+
+  /**
+   * Write {@link RunRecordMeta} to heart beat dataset as value.
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatTable.java]: javadoc: `table`

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 5]
```diff
@@ -769,4 +772,36 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
...
+      .withId(PROGRAM_HEARTBEATS)
+      .withFields(Fields.stringType(NAMESPACE_FIELD),
+                  Fields.longType(TIMESTAMP_FIELD),
```
Comment (CuriousVini) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: the name should include that timestamp is in secs.

File: cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatTable.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      Fields.stringField(StoreDefinition.ProgramHeartbeatStore.APPLICATION_FIELD, programRunId.getApplication()));
+    fields.add(
+      Fields.stringField(StoreDefinition.ProgramHeartbeatStore.PROGRAM_TYPE_FIELD, programRunId.getType().name()));
+    fields.add(Fields.stringField(StoreDefinition.ProgramHeartbeatStore.PROGRAM_FIELD, programRunId.getProgram()));
+    fields.add(Fields.stringField(StoreDefinition.ProgramHeartbeatStore.RUN_FIELD, programRunId.getRun()));
+    return fields;
+  }
+
+  /**
+   * Add namespace and timestamp and return it as the scan key
```
Comment (CuriousVini) [inline cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatTable.java]: javadoc: period at the end

**Review by CuriousVini — COMMENTED** [review]
A few comments and a question about TTL

File: cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java (hunk 3 above)
Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java:78]: hmm that's a good point - I didn't see this. It won't - so if needed we need to be able to do this in the SPI somehow

Comment (jayfjin) [reply cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java:78]: discussed with albert, added todo

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 2 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: copy paste :p

**Comment by jayfjin** [comment]
Addressed comments, PTAL

File: cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java:78]: sg

**Review by CuriousVini — APPROVED** [review]
lgtm 👍

**Comment by CuriousVini** [comment]
Please run ITM and DUT before merging.

**Comment by jayfjin** [comment]
IUT and DUT already passed, only changes made were comments/renaming

**Pull request merged**